src/sc/sidechainTxsCommitmentBuilder.cpp

```diff
@@ -0,0 +1,177 @@
+
+        std::vector<field_t*> singleSidechainComponents = {ftRootField, btrRootField, certRootField, scIdField};
+        field_t * sidechainTreeRoot = merkleTreeRootOf(singleSidechainComponents);
+
+        vSortedScLeaves.push_back(sidechainTreeRoot);
+    }
+
+    field_t * finalTreeRoot = emptyField.zero;
+    if (vSortedScLeaves.size() != 0)
+        finalTreeRoot = merkleTreeRootOf(vSortedScLeaves);
+
+    uint256 res = mapFieldToHash(finalTreeRoot);
+    if (finalTreeRoot != emptyField.zero) zendoo_field_free(finalTreeRoot);
+
+    return res;
+}
+#endif
+
+field_t* SidechainTxsCommitmentBuilder::mapScTxToField(const uint256& ccoutHash, const uint256& txHash, unsigned int outPos) const
+{
+    static_assert((sizeof(uint256) + sizeof(uint256) + sizeof(unsigned int)) < SC_FIELD_SIZE);
+
+    std::string toHash = std::string(ccoutHash.begin(), ccoutHash.end()) +
+            std::string(txHash.begin(), txHash.end()) +
+            std::string(static_cast<char*>(static_cast<void*>(&outPos)));
+
+    unsigned char hash[SC_FIELD_SIZE] = {};
+    memcpy(hash, toHash.append(SC_FIELD_SIZE - toHash.size(), '\0').c_str(), SC_FIELD_SIZE);
+
+    //generate field element. MISSING null-check
+    field_t* field = zendoo_deserialize_field(hash);
+    return field;
+}
+
+field_t* SidechainTxsCommitmentBuilder::mapCertToField(const uint256& certHash) const
+{
+    static_assert(sizeof(uint256) < SC_FIELD_SIZE);
+    unsigned char hash[SC_FIELD_SIZE] = {};
+    std::string resHex = std::string(certHash.begin(), certHash.end());
+    memcpy(hash, resHex.append(SC_FIELD_SIZE - resHex.size(), '\0').c_str(), SC_FIELD_SIZE);
+
+    //generate field element. MISSING null-check
+    field_t* field = zendoo_deserialize_field(hash);
+    return field;
+}
+
+uint256 SidechainTxsCommitmentBuilder::mapFieldToHash(const field_t* pField) const
+{
+    if (pField == nullptr)
+        throw;
+
+    unsigned char field_bytes[SC_FIELD_SIZE];
+    zendoo_serialize_field(pField, field_bytes);
+
+    uint256 res = Hash(BEGIN(field_bytes), END(field_bytes));
+    return res;
+}
+
+unsigned int SidechainTxsCommitmentBuilder::treeHeightForLeaves(unsigned int numberOfLeaves) const
+{
+    if (numberOfLeaves == 1)
+        return 2;
+
+    unsigned int res = static_cast<unsigned int>(ceil(log2f(static_cast<float>(numberOfLeaves)))) + 1;
+    return res;
+}
+
+field_t* SidechainTxsCommitmentBuilder::merkleTreeRootOf(std::vector<field_t*>& leaves) const
+{
+    //Notes: leaves are consumed, i.e. freed and nulled, if non-empty
+    //       root is a deep-copy, hence it must be freed later on
+    //       null leaves are mapped to emptyField
+    //       it is guaranteed not to return nullptr
+    field_t* res = nullptr;
+    auto btrTree = ZendooGingerRandomAccessMerkleTree(treeHeightForLeaves(leaves.size()));
+    for(field_t* & leaf: leaves)
+    {
+        //As of now we swallow possible errors in deserialization of field by replacing null with empty field
```

Comment:
This is to be discussed. Are genuine deserialization error supposed to happen ? A malicious user is able to craft a valid transaction that would lead to such error ? If yes, do we skip the leaf or do we set it to zero ?